conan/api/model.py

```diff
@@ -32,107 +30,60 @@ def __repr__(self):
-        }
-
-
-class _PackageUploadData:
-    def __init__(self, pref):
-        self.pref = pref
-        self.upload = True
-        self.files = None
-        self.force = None
-
-    def serialize(self):
-        return {
-            "pref": repr(self.pref),
-            "upload": self.upload,
-            "force": self.force,
-            "files": self.files
-        }
-
-
 class SelectBundle:
     def __init__(self):
-        self.recipes = OrderedDict()
-        self.confs = {}
-
-    def add_configurations(self, confs):
-        self.confs = {PkgReference(k.ref, k.package_id): v for k, v in confs.items()}
+        self.recipes = {}
 
     def add_refs(self, refs):
+        # RREVS alreday come in ASCENDING order, so upload does older revisions first
         for ref in refs:
-            self.recipes.setdefault(ref, [])
-
-    def refs(self):
-        return self.recipes.keys()
-
-    def prefs(self):
-        prefs = []
-        for v in self.recipes.values():
-            prefs.extend(v)
-        return prefs
+            ref_dict = self.recipes.setdefault(str(ref), {})
+            if ref.revision:
+                revs_dict = ref_dict.setdefault("revisions", {})
+                rev_dict = revs_dict.setdefault(ref.revision, {})
+                if ref.timestamp:
+                    rev_dict["timestamp"] = ref.timestamp
 
     def add_prefs(self, prefs):
+        # Prevs already come in ASCENDING order, so upload does older revisions first
         for pref in prefs:
-            binary_info = self.confs.get(PkgReference(pref.ref, pref.package_id))
-            self.recipes.setdefault(pref.ref, []).append((pref, binary_info))
-
-    def serialize(self):
-        ret = {}
-        for ref, prefs in sorted(self.recipes.items()):
-            ref_dict = ret.setdefault(str(ref), {})
-            if ref.revision:
-                revisions_dict = ref_dict.setdefault("revisions", {})
-                rev_dict = revisions_dict.setdefault(ref.revision, {})
-                if ref.timestamp:
-                    rev_dict["timestamp"] = timestamp_to_str(ref.timestamp)
-                if prefs:
-                    packages_dict = rev_dict.setdefault("packages", {})
-                    for pref_info in prefs:
-                        pref, info = pref_info
-                        pid_dict = packages_dict.setdefault(pref.package_id, {})
-                        if info is not None:
-                            pid_dict["info"] = info
-                        if pref.revision:
-                            prevs_dict = pid_dict.setdefault("revisions", {})
-                            prev_dict = prevs_dict.setdefault(pref.revision, {})
-                            if pref.timestamp:
-                                prev_dict["timestamp"] = timestamp_to_str(pref.timestamp)
-        return ret
+            revs_dict = self.recipes[str(pref.ref)]["revisions"]
+            rev_dict = revs_dict[pref.ref.revision]
```

Comment:
why not `setdefault` here as well? may it raise `KeyError` in some circumstances?